presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcFileRewriter.java

```diff
@@ -0,0 +1,194 @@
+                throw new IOException("File has fewer rows than deletion vector");
+            }
+            int deleteRowCount = rowsToDelete.cardinality();
+            if (fileRowCount == deleteRowCount) {
+                return new OrcFileInfo(0, 0);
+            }
+            if (fileRowCount >= Integer.MAX_VALUE) {
+                throw new IOException("File has too many rows");
+            }
+
+            OrcFileInfo fileInfo;
+            Map<String, String> metadata = OrcFileMetadata.from(newChunkId, columnIds, columnTypes).toMap();
+            try (OrcWriter writer = createOrcFileWriter(output, columnIds, storageTypes, compressionType, metadata, stats)) {
+                fileInfo = rewrite(recordReader, writer, rowsToDelete, columnTypes);
+            }
+            log.debug("Rewrote file %s in %s (input rows: %s, output rows: %s)", input.getName(), nanosSince(start), fileRowCount, fileRowCount - deleteRowCount);
+            return fileInfo;
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to rewrite file: " + input, e);
+        }
+    }
+
+    private static OrcFileInfo rewrite(OrcRecordReader reader, OrcWriter writer, BitSet rowsToDelete, List<Type> types)
+            throws IOException
+    {
+        int row = rowsToDelete.nextClearBit(0);
+        long rowCount = 0;
+        long uncompressedSize = 0;
+
+        while (true) {
+            if (Thread.currentThread().isInterrupted()) {
+                throw new InterruptedIOException();
+            }
+
+            int batchSize = reader.nextBatch();
+            if (batchSize <= 0) {
+                break;
+            }
+
+            if (row >= (reader.getFilePosition() + batchSize)) {
+                continue;
+            }
+
+            Block[] blocks = new Block[types.size()];
+            for (int i = 0; i < types.size(); i++) {
+                blocks[i] = reader.readBlock(types.get(i), i);
+            }
+
+            row = toIntExact(reader.getFilePosition());
+
+            Page page = maskedPage(blocks, rowsToDelete, row, batchSize);
+            writer.write(page);
+
+            rowCount += page.getPositionCount();
+            uncompressedSize += page.getLogicalSizeInBytes();
+
+            row = rowsToDelete.nextClearBit(row + batchSize);
+        }
+
+        return new OrcFileInfo(rowCount, uncompressedSize);
+    }
+
+    private static Page maskedPage(Block[] blocks, BitSet rowsToDelete, int start, int count)
+    {
+        int[] ids = new int[count];
+        int size = 0;
+        for (int i = 0; i < count; i++) {
+            if (!rowsToDelete.get(start + i)) {
+                ids[size] = i;
+                size++;
+            }
+        }
+
+        Block[] maskedBlocks = new Block[blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            maskedBlocks[i] = new DictionaryBlock(size, blocks[i], ids);
+        }
+        return new Page(maskedBlocks);
```

Comment:
"highker on Feb 4  Member
I believe Page::getPositions is doing the same thing.

  @zhaotao1986
zhaotao1986 on Feb 14  Member
There is no Page in this context. This line is the first place that creates a Page.

  @highker
highker on Feb 20  Member
Rewrite the following code

Block[] maskedBlocks = new Block[blocks.length];
for (int i = 0; i < blocks.length; i++) {
    maskedBlocks[i] = new DictionaryBlock(size, blocks[i], ids);
}
return new Page(maskedBlocks);
into

return new Page(maskedBlocks).getPositions(id, ...);"


This is indeed simpler. But it took me quite a while to verify the "Page.getPositions" correctness, and how to pass arguments to be correct. Are you certain "Page.getPositions" is good itself? 
So I haven't changed yet, allow me to put this as the last change.

Reply:
I read Page.getPositions many times. I changed it, in latest commit. 
The arguments should be (ids, 0, size);  second arg is always 0,